### ISSUE TITLE
unable to install perl-5.14.4

### ISSUE THREAD

**Issue by amit4mchiba** [gugod/App-perlbrew#611]
Hi, 
I am writing here to serious seek your advice in installing perl-5.14.4 using perlbrew. I installed perlbrew as instructed and sourced it. I checked and can see it working as all commands seems doing what they are suppose to do. My computer has ubuntu 16.04 OS, and it has perl-5.22.1 . For one of tools, I had to install perl-5.14.4. So, i gave command- perlbrew install perl-5.14.4.....and then it downloaded perl from cran and tried installing but finally gave error, saying that few test failed and not installed. Here is what I got-
amit8chiba@amit8chiba-Precision-Tower-7910:~$ perlbrew install perl-5.14.2
Fetching perl 5.14.2 as /home/amit8chiba/perl5/perlbrew/dists/perl-5.14.2.tar.bz2
Download http://www.cpan.org/authors/id/F/FL/FLORA/perl-5.14.2.tar.bz2 to /home/amit8chiba/perl5/perlbrew/dists/perl-5.14.2.tar.bz2
Installing /home/amit8chiba/perl5/perlbrew/build/perl-5.14.2/perl-5.14.2 into ~/perl5/perlbrew/perls/perl-5.14.2

This could take a while. You can run the following command on another shell to track the status:

  tail -f ~/perl5/perlbrew/build.perl-5.14.2.log

Installation process failed. To spot any issues, check

  /home/amit8chiba/perl5/perlbrew/build.perl-5.14.2.log

If some perl tests failed and you still want to install this distribution anyway,
do:

  (cd /home/amit8chiba/perl5/perlbrew/build/perl-5.14.2/perl-5.14.2; make install)

You might also want to try upgrading patchperl before trying again:

  perlbrew install-patchperl

Generally, if you need to install a perl distribution known to have minor test
failures, do one of these command to avoid seeing this message

  perlbrew --notest install perl-5.14.2
  perlbrew --force install perl-5.14.2

I did upgraded patchperl, and tried installing again, but it failed again. So, i decided to do force install, which eventually worked and i got perl-5.14.2. Then i switched to this perl and tried running the program and got this error-Segmentation fault (core dumped). I then got to another system which has perl5.10.1 and tried running same script and got no error. So, i guess this forced installation did not worked and still have error that does not allow running my program.

I then tried installing cperl-5.27.2 and it was installed successfully. So, the installation error is mostly when i am trying to install 5.10.1, 5.14.2, 5.18.2 and so on. But when I tried installing perl-5.22 and above or cperl versions, it seems working fine. 

Please advise me how shall i take up this issue and resolve. I really do not want to replace my current perl version and wanted to use perlbrew to install correctly and use perl-5.14.2. Please let me know if you need any further details on my approach or error or system configuration that i have in order to help me. 

thank you so much,
Amit

**Comment by amit4mchiba** [comment]
here is the installation log file for your reference in case it helps u to guide me...thank you in advance.
[build.perl-5.14.2.log](https://github.com/gugod/App-perlbrew/files/1934647/build.perl-5.14.2.log)

**Comment by gugod** [comment]
@amit4mchiba the end of that log reveals the fact that a few tests had failed, but the beginning suggest that `patchperl` itself failed to patch (probably):

```
Auto-guessed '5.14.2'
patching Configure
File Configure is read-only; trying to patch anyway
patching ext/Errno/Errno_pm.PL
File ext/Errno/Errno_pm.PL is read-only; trying to patch anyway
patching make_ext.pl
File make_ext.pl is read-only; trying to patch anyway
```


It mentiond that the file `Configure`, `Error_pm.PL` are read-only. Those lines are messages from `patch` program, and I doubt if ti became successful in the later attempts.

I suggest you to look into why that is the case. Assuming that you file system is not read-only, perhaps the way perlbrew extract the tarball generate a few read-only files. If you can still re-produce these messages, doing `ls -l` on those files and some of other files (for reference) will also be helpful.

**Comment by b33rcity** [comment]
I'm having a similar issue with 5.8.8. I tried going through the files listed and running `chmod +x` against each, but it seems that the install process does a download/extract each time and gives each file 444 permissions and negates that simple workaround.

**Comment by vsespb** [comment]
Same issue with 5.18 (`perlbrew install  perl-5.18.4 --multi --thread`)
my perlbrew version:
```
perlbrew --version
/home/vse/perl5/perlbrew/bin/perlbrew  - App::perlbrew/0.84
```
I have Debian 9. My home permissions seems ok.


/home perms:
```
drwxr-xr-x  11 root root  4096 май 13 02:21 home
```

/home/vse perms:

```
drwxr-xr-x 50 vse        vse      4096 июн 24 11:34 vse
```

my patchperl upgraded today.
```
vse@GREEN-U:~$ grep Devel::PatchPerl::VERSION `which patchperl`
  $Devel::PatchPerl::VERSION = '1.52';
    my $dpv = $Devel::PatchPerl::VERSION || "(unreleased)";
    my $dpv = $Devel::PatchPerl::VERSION || "(unreleased)";
```

I tried to workaround by pathcing FatPacker'ed version with `'chmod -R +w *'`:

```
      my @preconfigure_commands = (
          "cd $dist_extracted_dir",
+	  'chmod -R +w *',
          "rm -f config.sh Policy.sh",
      );
      push @preconfigure_commands, $patchperl unless $self->{"no-patchperl"} || $looks_like_we_are_installing_cperl;
```


Message "File Configure is read-only; trying to patch anyway" gone after that:

```
Auto-guessed '5.18.4'
patching Configure
patching ext/Errno/Errno_pm.PL
 
Beginning of configuration questions for perl5.
 
Checking echo to see how to suppress newlines...
...using \c
The star should be here-->*
 
First let's make sure your kit is complete.  Checking...
Looks good...

```

but I still have failures like perl not patched

```
../lib/filetest.t ................................................. ok
# Failed test 5 - preamble compiles at ../lib/h2ph.t line 49
#      got 'Bareword found where operator expected at _h2ph_pre.ph line 222, near "7fff"
# 	(Missing operator before fff?)
# Bareword found where operator expected at _h2ph_pre.ph line 228, near "7fffffff"
# 	(Missing operator before fffffff?)
# Bareword found where operator expected at _h2ph_pre.ph line 234, near "7fffffffffffffffL"
# 	(Missing operator before fffffffffffffffL?)
# Bareword found where operator expected at _h2ph_pre.ph line 240, near "7f"
# 	(Missing operator before f?)
# Bareword found where operator expected at _h2ph_pre.ph line 246, near "7fffffffffffffffL"
# 	(Missing operator before fffffffffffffffL?)
# Bareword found where operator expected at _h2ph_pre.ph line 250, near "7fffffffffffffffL"
# 	(Missing operator before fffffffffffffffL?)
# Bareword found where operator expected at _h2ph_pre.ph line 254, near "7fffffffffffffffL"
# 	(Missing operator before fffffffffffffffL?)
# Bareword found where operator expected at _h2ph_pre.ph line 258, near "7fffffffffffffffL"
# 	(Missing operator before fffffffffffffffL?)
# Bareword found where operator expected at _h2ph_pre.ph line 262, near "7fffffffffffffffL"
# 	(Missing operator before fffffffffffffffL?)
# Bareword found where operator expected at _h2ph_pre.ph line 266, near "7f"
# 	(Missing operator before f?)
# syntax error at _h2ph_pre.ph line 222, near "7fff "
# syntax error at _h2ph_pre.ph line 228, near "7fffffff "
# syntax error at _h2ph_pre.ph line 234, near "7fffffffffffffffL "
# syntax error at _h2ph_pre.ph line 240, near "7f "
# syntax error at _h2ph_pre.ph line 246, near "7fffffffffffffffL "
# syntax error at _h2ph_pre.ph line 250, near "7fffffffffffffffL "
# syntax error at _h2ph_pre.ph line 254, near "7fffffffffffffffL "
# syntax error at _h2ph_pre.ph line 258, near "7fffffffffffffffL "
# syntax error at _h2ph_pre.ph line 262, near "7fffffffffffffffL "
# syntax error at _h2ph_pre.ph line 266, near "7f "
# _h2ph_pre.ph has too many errors.
# '
# expected /(?^:syntax OK$)/
# Failed test 6 - output free of warnings at ../lib/h2ph.t line 56
#      got "Bareword found where operator expected at _h2ph_pre.ph line 222, near \"7fff\"\nCompilation failed in require at lib/h2ph.pht line 1.\nCompilation failed in require at -e line 1.\n"
# expected ""
../lib/h2ph.t ..................................................... 
Failed 2/6 subtests 

```

also there are other failures like:
```
# '-9223372036854775808' ne '-9.22337203685478e+18',    -1.84467440737096e+19   => I - N P vs I - P
```


Also, I think I found existing related issues:

https://github.com/tokuhirom/plenv/issues/116
https://github.com/bingos/devel-patchperl/pull/25


Attaching log with unaltered perlbrew:

[perlbrew-unaltered.log](https://github.com/gugod/App-perlbrew/files/2130664/perlbrew-unaltered.log)

and log with perlbrew altered with `'chmod -R +w *'`:

[perlbrew-patched.log](https://github.com/gugod/App-perlbrew/files/2130665/perlbrew-patched.log)

**Comment by vsespb** [comment]
so it's broken for perl-5.18 and below, thus unusable for me

**Comment by scottchiefbaker** [comment]
I'm having the same `h2ph` failure that you're getting here. I have a vanilla Fedora 28 VM, with `patchperl` installed and I'm unable to compile anything prior to 5.20. 

Versions 5.20 through 5.29 all compile and install just fine, but older Perls are not working.

They all complain about `Bareword found where operator expected at _h2ph_pre.ph`

**Comment by scottchiefbaker** [comment]
For what it's worth I just built a vanilla CentOS 7 VM and did the exact same process and 5.18 was able to compile just fine. 5.16 failed with the same `bareword` error as above though. Very odd.

Perhaps it's just an issues with newer build tools?

**Comment by scottchiefbaker** [comment]
I just installed a vanilla CentOS 7 VM. Using the same procedure I used on Fedora 28 I'm able to get 5.18 and 5.16 to install no problem. 

Perhaps it's just an issue with newer build tools?

**Comment by mackdoughe** [comment]
The issue with `../lib/h2ph.t` stems from newer versions of GCC being used with to compile older versions of Perl.

The ultimate fix for this involves patchperl being updated to apply the necessary patches that were incorporated as part of this [Bugreport 123784](https://rt.perl.org/Public/Bug/Display.html?id=123784). This would then allow perlbrew to successfully install these older Perl versions with newer build tools. See my more in-depth comment at: https://github.com/bingos/devel-patchperl/issues/23

**Comment by tcdaly** [comment]
I am having a similar problem compiling Perl 5.14.4 on Debian 10.  I'm getting 2 time-related compile failures as follows:

```#   Failed test 'timelocal year for 1970 1 2 0 0 0'
#   at t/Local.t line 108.
#          got: '170'
#     expected: '70'

#   Failed test 'timegm year for 1970 1 2 0 0 0'
#   at t/Local.t line 124.
#          got: '170'
#     expected: '70'
# Looks like you failed 2 tests of 163.
../cpan/Time-Local/t/Local.t ......................................
Dubious, test returned 2 (wstat 512, 0x200)
Failed 2/163 subtests
```

And 2 `../lib/h2ph.t` related failures:

```# Failed at ../lib/h2ph.t line 49
#      got 'Illegal declaration of subroutine main::__has_include at _h2ph_pre.ph line 687.
# '
# expected /(?^:syntax OK$)/
# Failed at ../lib/h2ph.t line 56
#      got "Illegal character in prototype for main::__has_include : STR at _h2ph_pre.ph line 687.\nCompilation fai
led in require at lib/h2ph.pht line 1.\nCompilation failed in require at -e line 1.\n"
# expected ""
../lib/h2ph.t .....................................................
Failed 2/6 subtests
```

**Comment by ufobat** [comment]
similar problem here:
```
../lib/feature/unicode_strings.t .................................. ok
../lib/filetest.t ................................................. ok
# Failed test 5 - preamble compiles at ../lib/h2ph.t line 49
#      got 'Illegal declaration of subroutine main::__has_include at _h2ph_pre.ph line 691.
# '
# expected /(?^:syntax OK$)/
# Failed test 6 - output free of warnings at ../lib/h2ph.t line 56
#      got "Illegal character in prototype for main::__has_include : STR at _h2ph_pre.ph line 691.\nCompilation failed in require at lib/h2ph.pht line 1.\nCompilation failed in require at -e line 1.\n"
# expected ""
../lib/h2ph.t ..................................................... 
Failed 2/6 subtests 
../lib/h2xs.t ..................................................... ok
```

**Comment by knmiecc** [comment]
Following the issue described [here](https://github.com/Perl/perl5/issues/14491) and the reference of a colleague [here](https://github.com/Perl/perl5/commit/6f87f404fa51739971a4068da1f11443024f3fc4) we were able to pinpoint the issue to a committed fix [here](https://github.com/Perl/perl5/commit/6f87f404fa51739971a4068da1f11443024f3fc4). 
**TLDR**: You can fix the Configure script manually. See below.

We were able to manually "patch" the Configure script generated by perlbrew and rerun the installation of the desired version. 
However, because we have frankly no idea how patches are applied with patchperl and do not know where to propose any changes in that regard anyway, I hope the manual fix below might help others encountering this issue.
Also if someone can point us or the fix in the right direction that would also help.

**To recreate the manual fix**:
Let perlbrew try to install 5.14.4 and have it fail by running _perlbrew install perl-5.14.4_. 
Change into the build directory for that version, in our case it was `/opt/perl5/perlbrew/build/perl-5.14.4./perl-5.14.4/`. Change into the directory and open the Configure script file and replace the respective line marked with a minus, and replace with the line below marked with a plus. The exact line number differs from your specific set of compile flags. Please [see this commit ](https://github.com/Perl/perl5/commit/6f87f404fa51739971a4068da1f11443024f3fc4) to get a better git diff like overview of the change.
```bash
.....
- ./Cppsym < Cppsym.know > Cppsym.true
+ ./Cppsym < Cppsym.know | $sort | $uniq > Cppsym.true
.....
- sed 's/#define[\ \  ]*//;s/[\ \     ].*$//' <$tHdrH'_cppsym.h' >$tHdrH'_cppsym.real'
+ sed -e 's/#define[\ \  ]*//;s/[\ \     ].*$//' -e 's/(.*//' <$tHdrH'_cppsym.h' >$tHdrH'_cppsym.real'
....
```

Save that and rerun the installation, that is otherwise done automatically by perlbrew. You can omit ${PERLBREW_CONFIGURE_FLAGS} if you set your configuration differently.
```bash
sh ./Configure ${PERLBREW_CONFIGURE_FLAGS}
make
make test_harness
make install
```
